contracts/AelinUpFrontDeal.sol

```diff
@@ -0,0 +1,357 @@
+        UpFrontDeal memory _dealData = dealData;
+
+        require(
+            IERC20(_dealData.underlyingDealToken).balanceOf(msg.sender) >= _depositUnderlyingAmount,
+            "not enough balance"
+        );
+        require(!underlyingDepositComplete, "already deposited the total");
+
+        uint256 balanceBeforeTransfer = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        IERC20(_dealData.underlyingDealToken).transferFrom(msg.sender, address(this), _depositUnderlyingAmount);
+        uint256 balanceAfterTransfer = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        uint256 underlyingDealTokenAmount = balanceAfterTransfer - balanceBeforeTransfer;
+
+        if (balanceAfterTransfer >= _dealData.underlyingDealTokenTotal) {
+            underlyingDepositComplete = true;
+            purchaseExpiry = block.timestamp + _dealData.purchaseDuration;
+            emit DealFullyFunded(address(this), block.timestamp, purchaseExpiry);
+        }
+
+        emit DepositDealToken(_dealData.underlyingDealToken, msg.sender, underlyingDealTokenAmount);
+    }
+
+    /**
+     * @dev allows holder to withdraw any excess underlying deal tokens deposited to the contract
+     */
+    function withdrawExcess() external onlyHolder {
+        UpFrontDeal memory _dealData = dealData;
+        uint256 currentBalance = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+        require(currentBalance > _dealData.underlyingDealTokenTotal, "no excess to withdraw");
+
+        uint256 excessAmount = currentBalance - _dealData.underlyingDealTokenTotal;
+        IERC20(_dealData.underlyingDealToken).transferFrom(address(this), msg.sender, excessAmount);
+        currentBalance = IERC20(_dealData.underlyingDealToken).balanceOf(address(this));
+
+        require(currentBalance == _dealData.underlyingDealTokenTotal, "incorrect balance");
+        emit WithdrewExcess(address(this), excessAmount);
+    }
+
+    function acceptDeal(IAelinPool.NftPurchaseList[] calldata _nftPurchaseList, uint256 _purchaseTokenAmount) external {
+        UpFrontDeal memory _dealData = dealData;
+        require(underlyingDepositComplete, "deal token not yet deposited");
+        require(!poolDepositComplete, "purchase token cap reached");
+        require(block.timestamp < purchaseExpiry, "not in purchase window");
+        require(IERC20(_dealData.purchaseToken).balanceOf(msg.sender) >= _purchaseTokenAmount, "not enough purchaseToken");
+        uint256 purchaseTokenAmount;
+
+        if (_nftPurchaseList.length > 0) {
+            purchaseTokenAmount = _purchaseDealTokensWithNft(_dealData, _nftPurchaseList, _purchaseTokenAmount);
+        } else {
+            purchaseTokenAmount = _purchaseDealTokens(_dealData, _purchaseTokenAmount);
+        }
+
+        totalAmountAccepted += purchaseTokenAmount;
+        amountPurchaseTokens[msg.sender] += purchaseTokenAmount;
```

Comment:
can we name `amountPurchaseTokens` something else? maybe `purchaseTokensPerAccount` or `purchaseTokenPerUser`